File: potential_territory.rb
require_relative "./board_analyser"

class PotentialTerritory

  def initialize(goban)
    @goban = goban
    @size = goban.size
    @boan = BoardAnalyser.new
    @real_grid = @goban.scoring_grid # we can reuse the already allocated grid
    @real_yx = @real_grid.yx # simple shortcut to real yx
    # grids below are used in the evaluation process
    @grids = [Grid.new(goban.size), Grid.new(goban.size)]
    @reduced_grid = Grid.new(goban.size)
    @territory = Grid.new(goban.size) # result of evaluation
  end

  # Returns the matrix of potential territory.
  # +1: definitely white, -1: definitely black
  # Values in between are possible too.
  def guess_territories
    # update real grid to current goban
    @real_grid.convert(@goban.grid)
    # evaluate 2 "scenarios" - each player plays everywhere *first*
    2.times {|first| foresee(@grids[first], first, 1-first)}
    $log.debug("\nBLACK first:\n#{@grids[0]}WHITE first:\n#{@grids[1]}") if $debug

    # now merge the result
    1.upto(@size) do |j|
      1.upto(@size) do |i|
        owner = 0
        2.times do |first|
          terr_color = @grids[first].yx[j][i] - Grid::TERRITORY_COLOR
          owner += 1 if terr_color == WHITE
          owner -= 1 if terr_color == BLACK
        end
        @territory.yx[j][i] = owner / 2.0
      end
    end
    $log.debug("\n+1=white, -1=black, 0=no one\n"+@territory.to_text{|v| v==0 ? "    0" : sprintf("%+.1f",v)}) if $debug
    return @territory.yx
  end

  def potential
    return @territory
  end

  # For unit tests
  def _grid(first)
    return @grids[first]
  end

private

  # TODO: add live/dead groups? Maybe not here
  def foresee(grid, first, second)
    @tmp = @territory # safe to use it as temp grid here
    @reduced_yx = nil
    @move_num_before_enlarge = @goban.move_number?

    # enlarging starts with real grid
    enlarge(@real_grid, @tmp.copy(@real_grid), first, second)
    enlarge(@tmp, grid.copy(@tmp), second, first)
    connect_to_borders(grid.yx)
    $log.debug("after 1st enlarge:\n#{@grid}") if $debug

    # for reducing we start from the enlarged grid
    reduce(@reduced_grid.copy(grid))
    @reduced_yx = @reduced_grid.yx
    $log.debug("after reduce:\n#{grid}") if $debug

    # now we have the reduced goban, play the enlarge moves again minus the extra
    enlarge(@real_grid, @tmp.copy(@real_grid), first, second)
    enlarge(@tmp, grid.copy(@tmp), second, first)
    connect_to_borders(grid.yx)
    $log.debug("after 2nd enlarge:") if $debug
    @goban.debug_display if $debug

    # passed grid will receive the result (scoring grid)
    @boan.count_score(@goban, grid.convert(@goban.grid))

    # restore goban
    (@goban.move_number? - @move_num_before_enlarge).times { Stone.undo(@goban) }
  end

  def enlarge(in_grid, out_grid, first, second)
    $log.debug("enlarge #{first},#{second}") if $debug
    in_yx = in_grid.yx
    out_yx = out_grid.yx
    1.upto(@size) do |j|
      1.upto(@size) do |i|
        next if in_yx[j][i] != EMPTY
        enlarge_at(in_yx, out_yx, i, j, first, second)
      end
    end
  end

  # Reduces given grid using the real grid as reference.
  def reduce(grid)
    yx = grid.yx
    1.upto(@size) do |j|
      1.upto(@size) do |i|
        next if @real_yx[j][i] != EMPTY # cannot reduce a real stone
        color = yx[j][i]
        next if color == EMPTY # we did not enlarge here, no need to reduce
        enemies = in_contact?(yx, i, j, 1-color)
        # we can safely reduce if no enemy was around at the end of the enlarging steps
        yx[j][i] = EMPTY if enemies == 0
      end
    end
  end

  # "enlarge" around a given spot
  # Note we read and write on separate grids
  def enlarge_at(in_yx, out_yx, i, j, first, second)
    ss = in_contact?(in_yx, i, j, first)
    if ss > 0
      return if ss >= 3 # if 3 or 4 no need to fill the void
    else
      return if !diagonal_move_ok?(in_yx, i, j, first, second)
    end
    add_stone(out_yx, i, j, first)
  end

  # Add a stone on given grid.
  # When the reduced grid is known, use it and play moves on goban too.
  def add_stone(yx, i, j, color)
    if @reduced_yx
      # skip if useless move (it was reduced)
      return if @reduced_yx[j][i]==EMPTY
      # we check only against sucicide (e.g. no need to check against ko or non empty)
      stone = @goban.stone_at?(i,j)
      return if stone.move_is_suicide?(color)
      Stone.play_at(@goban, i, j, color)
    end
    yx[j][i] = color
  end

  # Returns the number of times we find "color" in contact with i,j
  def in_contact?(yx, i, j, color)
    num = 0
    Stone::XY_AROUND.each { |vect| num += 1 if yx[j+vect[1]][i+vect[0]] == color }
    return num
  end

  # Authorises a diagonal move if first color is on a diagonal stone from i,j
  # AND if second color is not next to this diagonal stone
  def diagonal_move_ok?(yx, i, j, first, second)
    Stone::XY_DIAGONAL.each do |vect|
      next if yx[j+vect[1]][i+vect[0]] != first
      next if yx[j+vect[1]][i] == second or yx[j][i+vect[0]] == second
      $log.debug("diagonal_move_ok: #{i},#{j} for #{first}") if $debug and i==1 and j==9
      return true
    end
    return false
  end

  AROUND = [[1,0,0,1],[0,1,1,0],[1,0,-1,0],[-1,0,1,0]] #TODO replace this by pre-computed coords

  # connect stones close to borders to the border
  def connect_to_borders(yx)
    2.upto(@size-1) do |n|
      AROUND.each do |c|
        i = (c[0]<0 ? @size   : c[0]*n)     + c[1] # n,1,n,size
        j = (c[2]<0 ? @size   : c[2]*n)     + c[3] # 1,n,size,n
        if yx[j][i]==EMPTY
          i2= (c[0]<0 ? @size-1 : c[0]*n)     + c[1]*2  # n,2,n,size-1
          j2= (c[2]<0 ? @size-1 : c[2]*n)     + c[3]*2  # 2,n,size-1,n
          i3= (c[0]<0 ? @size   : c[0]*(n+1)) + c[1]  # n+1,1,n+1,size
          j3= (c[2]<0 ? @size   : c[2]*(n+1)) + c[3]  # 1,n+1,size,n+1
          i4= (c[0]<0 ? @size   : c[0]*(n-1)) + c[1]  # n-1,1,n-1,size
          j4= (c[2]<0 ? @size   : c[2]*(n-1)) + c[3]  # 1,n-1,size,n-1
          next2border = yx[j2][i2]
          if next2border!=EMPTY and yx[j3][i3]==EMPTY and yx[j4][i4]==EMPTY
            add_stone(yx, i, j, next2border)
          end
        end
      end
      # if @goban.empty?(i,1)
      #   next2border = @goban.stone_at?(i,2).color
      #   if next2border != EMPTY and @goban.empty?(i+1,1) and @goban.empty?(i-1,1)
      #     Stone.play_at(@goban,i,1,next2border)
      #   end
      # end
      # if @goban.empty?(1,i)
      #   next2border = @goban.stone_at?(2,i).color
      #   if next2border != EMPTY and @goban.empty?(1,i+1) and @goban.empty?(1,i-1)
      #     Stone.play_at(@goban,1,i,next2border)
      #   end
      # end
      # if @goban.empty?(i,@size)
      #   next2border = @goban.stone_at?(i,@size-1).color
      #   if next2border != EMPTY and @goban.empty?(i+1,@size) and @goban.empty?(i-1,@size)
      #     Stone.play_at(@goban,i,@size,next2border)
      #   end
      # end
      # if @goban.empty?(@size,i)
      #   next2border = @goban.stone_at?(@size-1,i).color
      #   if next2border != EMPTY and @goban.empty?(@size,i+1) and @goban.empty?(@size,i-1)
      #     Stone.play_at(@goban,@size,i,next2border)
      #   end
      # end
    end
  end

end

File: js/PotentialTerritory.js
//Translated from potential_territory.rb using babyruby2js
'use strict';

var Grid = require('./Grid');
var main = require('./main');
var Stone = require('./Stone');
var BoardAnalyser = require('./BoardAnalyser');

var EMPTY = main.EMPTY, BLACK = main.BLACK, WHITE = main.WHITE;
var NEVER = main.NEVER;

var POT2CHAR = Grid.territory2char;
var POT2OWNER = Grid.territory2owner;

var UP = 0, RIGHT = 1, DOWN = 2, LEFT = 3;
var XY_AROUND = Stone.XY_AROUND; // UP RIGHT DOWN LEFT
var XY_DIAGONAL = Stone.XY_DIAGONAL;


/** @class */
function PotentialTerritory(goban) {
    this.goban = goban;
    this.gsize = goban.gsize;
    this.boan = new BoardAnalyser();
    this.realGrid = this.goban.scoringGrid; // we can reuse the already allocated grid
    this.realYx = this.realGrid.yx; // simple shortcut to real yx
    // grids below are used in the evaluation process
    this.grids = [new Grid(this.gsize), new Grid(this.gsize)];
    this.reducedGrid = new Grid(this.gsize);
    this.territory = new Grid(this.gsize); // result of evaluation
    this._prepareBorderConnect();
}
module.exports = PotentialTerritory;

// Returns the matrix of potential territory.
// +1: definitely white, -1: definitely black
// Values in between are possible too.
PotentialTerritory.prototype.guessTerritories = function () {
    this._initGroupState();
    // update real grid to current goban
    this.realGrid.initFromGoban(this.goban);
    // evaluate 2 "scenarios" - each player plays everywhere *first*
    this._foresee(this.grids[BLACK], BLACK, WHITE);
    this._foresee(this.grids[WHITE], WHITE, BLACK);
    // now merge the result
    var blackYx = this.grids[BLACK].yx;
    var whiteYx = this.grids[WHITE].yx;
    var resultYx = this.territory.yx;
    for (var j = 1; j <= this.gsize; j++) {
        for (var i = 1; i <= this.gsize; i++) {
            resultYx[j][i] = (POT2OWNER[2 + blackYx[j][i]] + POT2OWNER[2 + whiteYx[j][i]]) / 2;
        }
    }
    if (main.debug) main.log.debug('Guessing territory for:\n' + this.realGrid +
        '\nBLACK first:\n' + this.grids[BLACK] + 'WHITE first:\n' + this.grids[WHITE] + this);
    return resultYx;
};

PotentialTerritory.prototype.potential = function () {
    return this.territory;
};

PotentialTerritory.prototype.toString = function () {
    return '\n+1=white, -1=black, 0=no one\n' +
        this.territory.toText(function (v) { return v === 0 ? '    0' : '%+.1f'.format(v); }) +
        this.territory.toText(function (v) { return POT2CHAR[2 + v * 2]; });
};

PotentialTerritory.prototype.image = function () {
    return this.territory.toLine(function (v) { return POT2CHAR[2 + v * 2]; });
};

// For unit tests
PotentialTerritory.prototype._grid = function (first) {
    return this.grids[first];
};

PotentialTerritory.prototype._foresee = function (grid, first, second) {
    var moveCount = this.goban.moveNumber();

    // passed grid will receive the result (scoring grid)
    this._connectThings(grid, first, second);
    this.boan.analyse(first, this.goban, grid.initFromGoban(this.goban));
    this._collectGroupState();

    // restore goban
    moveCount = this.goban.moveNumber() - moveCount;
    while (moveCount-- > 0) Stone.undo(this.goban);
};

PotentialTerritory.prototype._initGroupState = function () {
    this.allGroups = this.goban.getAllGroups();
    for (var ndx in this.allGroups) {
        var g = this.allGroups[ndx];
        g.isAlive = g.isDead = NEVER;
    }
};

PotentialTerritory.prototype._collectGroupState = function () {
    for (var ndx in this.allGroups) {
        var g0 = this.allGroups[ndx], gn = g0;
        // follow merge history to get final group g0 ended up into
        while (gn.mergedWith) gn = gn.mergedWith;
        // collect state of final group
        if (gn.killedBy || gn._info.isDead) {
            g0.isDead++;
        } else if (gn._info.isAlive) {
            g0.isAlive++;
        }
    }
};

PotentialTerritory.prototype._connectThings = function (grid, first, second) {
    this.tmp = this.territory; // safe to use it as temp grid here
    this.reducedYx = null;
    // enlarging starts with real grid
    this.enlarge(this.realGrid, this.tmp.copy(this.realGrid), first, second);
    this.enlarge(this.tmp, grid.copy(this.tmp), second, first);
    if (main.debug) main.log.debug('after 1st enlarge (before connectToBorders):\n' + grid);
    this.connectToBorders(grid.yx);
    if (main.debug) main.log.debug('after connectToBorders:\n' + grid);

    // for reducing we start from the enlarged grid
    this.reduce(this.reducedGrid.copy(grid));
    this.reducedYx = this.reducedGrid.yx;
    if (main.debug) main.log.debug('after reduce:\n' + this.reducedGrid);

    // now we have the reduced goban, play the enlarge moves again minus the extra
    this.enlarge(this.realGrid, this.tmp.copy(this.realGrid), first, second);
    this.enlarge(this.tmp, grid.copy(this.tmp), second, first);
    if (main.debug) main.log.debug('after 2nd enlarge (before connectToBorders):\n' + grid);
    this.connectToBorders(grid.yx);
    if (main.debug) main.log.debug('after connectToBorders:\n' + grid);
};

PotentialTerritory.prototype.enlarge = function (inGrid, outGrid, first, second) {
    if (main.debug) main.log.debug('enlarge ' + first + ',' + second);
    var inYx = inGrid.yx;
    var outYx = outGrid.yx;
    for (var j = 1; j <= this.gsize; j++) {
        for (var i = 1; i <= this.gsize; i++) {
            if (inYx[j][i] !== EMPTY) {
                continue;
            }
            this.enlargeAt(inYx, outYx, i, j, first, second);
        }
    }
};

// Reduces given grid using the real grid as reference.
// We can safely "reduce" if no enemy was around at the end of the enlarging steps.
PotentialTerritory.prototype.reduce = function (grid) {
    var yx = grid.yx;
    for (var j = 1; j <= this.gsize; j++) {
        for (var i = 1; i <= this.gsize; i++) {
            // cannot reduce a real stone
            if (this.realYx[j][i] !== EMPTY) continue;
            
            var color = yx[j][i];
            // if we did not enlarge here, no need to reduce
            if (color === EMPTY) continue;

            // reduce if no enemy was around
            var enemies = this.inContact(yx, i, j, 1 - color);
            if (enemies === 0) {
                yx[j][i] = EMPTY;
            }
        }
    }
};

// "enlarge" around a given spot
// Note we read and write on separate grids
PotentialTerritory.prototype.enlargeAt = function (inYx, outYx, i, j, first, second) {
    switch (this.inContact(inYx, i, j, first)) {
    case 0: // no ally in line, check diagonal too
        if (!this.diagonalMoveOk(inYx, i, j, first, second)) return;
        break;
    case 4: // no need to fill the void
        return;
    case 3: // fill only if enemy is around
        if (this.inContact(inYx, i, j, second) === 0) return;
    }
    return this.addStone(outYx, i, j, first);
};

// Add a stone on given grid.
// When the reduced grid is known, use it and play moves on goban too.
PotentialTerritory.prototype.addStone = function (yx, i, j, color, border) {
    if (this.reducedYx) {
        // skip if useless move (it was reduced)
        if (!border && this.reducedYx[j][i] === EMPTY) {
            return;
        }
        // we check only against sucicide (e.g. no need to check against ko or non empty)
        var stone = this.goban.stoneAt(i, j);
        if (stone.moveIsSuicide(color)) {
            return;
        }
        Stone.playAt(this.goban, i, j, color);
    }
    yx[j][i] = color;
};

// Returns the number of times we find "color" in contact with i,j
PotentialTerritory.prototype.inContact = function (yx, i, j, color) {
    var num = 0;
    for (var dir = UP; dir <= LEFT; dir++) {
        var vect = XY_AROUND[dir];
        if (yx[j + vect[1]][i + vect[0]] === color) {
            num++;
        }
    }
    return num;
};

// Authorises a diagonal move if first color is on a diagonal stone from i,j
// AND if second color is not next to this diagonal stone
PotentialTerritory.prototype.diagonalMoveOk = function (yx, i, j, first, second) {
    for (var v = 0; v < 4; v++) {
        var vect = XY_DIAGONAL[v];
        if (yx[j + vect[1]][i + vect[0]] !== first) continue;
        if (yx[j + vect[1]][i] === second || yx[j][i + vect[0]] === second) continue;
        return true;
    }
    return false;
};

// Returns the position of specified border stone
function borderPos(dir, n, gsize) {
    switch (dir) {
    case UP: return [n, 1]; // bottom border
    case RIGHT: return [1, n]; // facing right => left border
    case DOWN: return [n, gsize]; // facing down => top border
    case LEFT: return [gsize, n]; // right border
    }
}

PotentialTerritory.prototype._prepareBorderConnect = function () {
    var points = [];
    for (var dir = UP; dir <= LEFT; dir++) {
        var dx = XY_AROUND[dir][0], dy = XY_AROUND[dir][1];
        for (var n = this.gsize - 1; n >= 2; n--) {
            var p = borderPos(dir, n, this.gsize);
            points.push([p[0], p[1], dx, dy]);
        }
    }
    this.borderPoints = points;
};

/** Example for left border: (direction = RIGHT)
 G=GOAL, S=SPOT, L=LEFT, R=RIGHT (left & right could be switched, it does not matter)
 +
 L0 L1 L2
 G  S1 S2 S3
 R0 R1 R2
 +
 */
PotentialTerritory.prototype.connectToBorders = function (yx) {
    var points = this.borderPoints;
    var l0, r0, l1, r1, l2, r2;
    for (var i = points.length - 1; i >=0; i--) {
        var p = points[i];
        var gi = p[0], gj = p[1];
        if (yx[gj][gi] !== EMPTY) continue;
        var dx = p[2], dy = p[3]; // direction from border to center
        var s1i = gi + dx, s1j = gj + dy; // spot 1
        var color = yx[s1j][s1i];
        if (color !== EMPTY) {
            l0 = yx[gj-dx][gi-dy]; r0 = yx[gj+dx][gi+dy];
            if (l0 === color || r0 === color) continue; // no need for goal if s0 or r0
            this.addStone(yx, gi, gj, color, true);
            continue;
        }
        var s2i = s1i + dx, s2j = s1j + dy;
        color = yx[s2j][s2i];
        if (color !== EMPTY) {
            l0 = yx[gj-dx][gi-dy]; r0 = yx[gj+dx][gi+dy];
            l1 = yx[s1j-dx][s1i-dy]; r1 = yx[s1j+dx][s1i+dy];
            if (l0 === color && l1 === color) continue;
            if (r0 === color && r1 === color) continue;
            this.addStone(yx, s1i, s1j, color, true);
            if (l0 === color || r0 === color) continue; // no need for goal if s0 or r0
            this.addStone(yx, gi, gj, color, true);
            continue;
        }
        var s3i = s2i + dx, s3j = s2j + dy;
        color = yx[s3j][s3i];
        if (color !== EMPTY) {
            l0 = yx[gj-dx][gi-dy]; r0 = yx[gj+dx][gi+dy];
            l1 = yx[s1j-dx][s1i-dy]; r1 = yx[s1j+dx][s1i+dy];
            l2 = yx[s2j-dx][s2i-dy]; r2 = yx[s2j+dx][s2i+dy];
            if (l0 === color && l1 === color && l2 === color) continue;
            if (r0 === color && r1 === color && r2 === color) continue;
            this.addStone(yx, s2i, s2j, color, true);
            if (l0 === color && l1 === color) continue;
            if (r0 === color && r1 === color) continue;
            this.addStone(yx, s1i, s1j, color, true);
            if (l0 === color || r0 === color) continue; // no need for goal if s0 or r0
            this.addStone(yx, gi, gj, color, true);
            continue;
        }
    }
};
